sklearn/metrics/_decision_threshold.py

```diff
@@ -0,0 +1,117 @@
+"""Metric per threshold curve to assess binary classification performance.
+
+Compute metric per threshold, over a range of threshold values to aid visualization
+of threshold-dependent metric behavior.
+
+Utilizes `_CurveScorer` methods to do all the computation.
+"""
+
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+from numbers import Integral
+
+from ..utils._param_validation import Interval, validate_params
+
+
+@validate_params(
+    {
+        "score_func": [callable],
+        "y_true": ["array-like"],
+        "y_score": ["array-like"],
+        "thresholds": [
+            Interval(Integral, 2, None, closed="left"),
+            "array-like",
+        ],
+        "greater_is_better": ["boolean"],
+        "labels": ["array-like", None],
+    },
+    prefer_skip_nested_validation=True,
+)
+def decision_threshold_curve(
+    score_func,
+    y_true,
+    y_score,
+    # Should below 2 have a default value?
+    thresholds=20,
+    greater_is_better=True,
+    labels=None,
+    **kwargs,
+):
+    """Compute threshold-dependent metric of interest per threshold.
+
+    Note: this implementation is restricted to the binary classification task.
+
+    Read more in the :ref:`User Guide <threshold_tunning>`.
+
+    .. versionadded:: 1.8
+
+    Parameters
+    ----------
+    score_func : callable
+        The score function to use. It will be called as
+        `score_func(y_true, y_pred, **kwargs)`.
+
+    y_true : array-like of shape (n_samples,)
+        Ground truth (correct) target labels.
+
+    y_score : array-like of shape (n_samples,)
+        Continuous response scores.
+
+    thresholds : int or array-like, default=20
+        Specifies number of decision thresholds to compute score for. If an integer,
+        it will be used to generate `thresholds` thresholds uniformly distributed
+        between the minimum and maximum of `y_score`. If an array-like, it will be
+        used as the thresholds.
```

Comment:
I believe we can compute the thresholds automatically, like in the other curve functions.
`score_func` is called on `y_true` (independent of the threshold) and `y_pred` where `y_pred` is computed from `y_score` and the threshold.

If we sort `y_score` in ascending order, it's only where the value of sorted_y_score changes that we need to set a threshold. That way we compute the exact number of thresholds. We can probably leverage some code from `_binary_clf_curve` https://github.com/scikit-learn/scikit-learn/blob/18e89a4fbfd742ac0af6316f13ccdc1a67f6e5af/sklearn/metrics/_ranking.py#L891-L906